README.md
# screen_ratio_adapter

A new Flutter package.Based on screen rectangle with wide orientation adaptation, height scrolling or leave blank. Looks just like the design. No more repetitive moving of bricks for adapting screens.
长得就像设计稿一样。再也不为适配屏幕重复搬砖。

设计稿如同模特上挂着的衣服，但是强行穿在不同设备上，就成了买家秀，要么左右不对称，甚至撑破。

<img src="https://ss0.bdstatic.com/70cFuHSh_Q1YnxGkpoWK1HF6hhy/it/u=420958367,1948311389&fm=26&gp=0.jpg" style="zoom:75%;" />

[TOC]

### 基于一种理念

一套多售，这是批量订制而不是私人定制。

服装设计稿——>加工厂批量生产不同尺码——>顾客选购——>着身；

中间环节由本插件代码处理

### 原理

还记得Android原生上[今日头条适配方案](https://juejin.im/post/5b7a29736fb9a019d53e7ee2)吗？就是修改运行时获取的设备屏幕密度比率常数。这个方案很早之前就有人提出过只是人微言轻传播的不好。

对于用户来说屏幕就是一个矩形（至少目前的主流设备）；而主流移动设备已经培养多年用户上下滑动屏幕的习惯；

主流设备变得更“窄”；

基于以上三点，以屏幕宽度适配通用性高，实现效果好。至于实际设备高/宽与设计比例的差异处理，自行留白或纵向改为scrollWidget

### 实践报告

**下图警戒区为故意超出一个单位的示例图**

| \设备                     | android                                                      | Android                                                      | iOS                                                          |
| ------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| 设备分辨率                | 1080x1794                                                    | 720x2160                                                     | 1242x2688                                                    |
| h/w                       | 1.666                                                        | 3.0                                                          | 2.164                                                        |
| 屏幕密度比率ρ             | 2.625                                                        | 1.2                                                          | 3.0                                                          |
|                           |                                                              |                                                              |                                                              |
| UI稿1（300x510) h/w =1.7  | 过长                                                         | 过短                                                         | 过短                                                         |
| 适配比率ρ(300x510)        | 3.60                                                         | 2.40                                                         | 4.14                                                         |
|                           | <img src="https://user-gold-cdn.xitu.io/2020/4/18/1718df72aa6ec4cf?w=1080&amp;h=1920&amp;f=png&amp;s=142475" style="zoom:25%;" /> | <img src="https://user-gold-cdn.xitu.io/2020/4/18/1718df4008003319?w=720&amp;h=2160&amp;f=png&amp;s=88085" style="zoom:25%;" /> | <img src="https://user-gold-cdn.xitu.io/2020/4/18/1718df3831bbbaa9?w=822&amp;h=1686&amp;f=png&amp;s=214200" style="zoom:25%;" /> |
| UI稿1（414x896) h/w =2.16 | 过长                                                         | 过短                                                         | 不变                                                         |
| 适配比率ρ(414x896)        | 2.608                                                        | 1.739                                                        | 3.0                                                          |
|                           | <img src="https://user-gold-cdn.xitu.io/2020/4/18/1718df4e1efb90dc?w=1080&amp;h=1920&amp;f=png&amp;s=148927" style="zoom:25%;" /> | <img src="https://user-gold-cdn.xitu.io/2020/4/18/1718df5c3119b784?w=720&amp;h=2160&amp;f=png&amp;s=90158" style="zoom:25%;" /> | <img src="https://user-gold-cdn.xitu.io/2020/4/18/1718df170d282fee?w=854&amp;h=1738&amp;f=png&amp;s=298566" style="zoom:25%;" /> |
|                           |                                                              |                                                              |                                                              |



[已上传**pub.dev** ️❤️❤️点赞😘😘](https://pub.dev/packages/screen_ratio_adapter)

[已上传**github**    ️❤️❤️加星😘😘](https://github.com/qcks/flutter_screen_ratio_adapter)

### 测试release模式下效果

```
Additional arguments:--release
```



### 使用方式

```yaml
点赞加星小礼物
复制粘贴人人爱
如有BUG记小本
携尔长刀来相砍
```


```yaml
dependencies:
  screen_ratio_adapter: ^latest version
```

```dart
import 'package:screen_ratio_adapter/screen_ratio_adapter.dart';
///设计稿尺寸，单位应是pt或dp
var uiSize = BlueprintsRectangle(414, 896);
//void main() => runApp(MyApp());
void main() {
  FxWidgetsFlutterBinding.ensureInitialized(
      uiBlueprints: uiSize);
  return runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(),
      builder: FxTransitionBuilder(builder: null),//勿忘
    );
  }
}
————————————————————————
  Container(
  child: Text("W= 415"),
  width: 415.w
  height: 80.h)；//❌
 Container(
  child: Text("W= 415"),
  width: 415
  height: 80)；//✅

```

### 参考及鸣谢

[一种粗暴快速的Android全屏幕适配方案](https://www.jianshu.com/p/b6b9bd1fba4d)

[flutter 屏幕适配](https://github.com/genius158/FlutterTest/blob/master/lib/main.dart)

[flutter屏幕适配 ，一种一劳永逸的全局适配方式](https://juejin.im/post/5cb49e306fb9a068a3729b41#heading-0)





CHANGELOG.md
## [0.0.1] - TODO: Add release date.
## [0.0.2] - TODO: 更新readme.
## [0.0.3] - TODO: fix bug.
## [0.1.0] - TODO: -bug(fix:Bug).
## [0.9.0] - TODO: 竖屏状态无干扰.
## [1.0.0] - TODO: To perform well.
## [1.0.1] - TODO: release bug fix.
## [1.0.2] - TODO: 代码自检升级.
## [2.0.0] - TODO: 空安全，Stable release for null safety.
## [2.0.1] - TODO: .
## [3.0.0] - TODO: 初始化方法简化，修改全局MediaQueryData.size,更新readme.

* TODO: Describe initial release.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:screen_ratio_adapter/screen_ratio_adapter.dart';

///设计稿尺寸，单位应是pt或dp
//var uiSize = BlueprintsRectangle(300, 510);
//var uiSize = BlueprintsRectangle(721, 628);
var uiSize = BlueprintsRectangle(414, 878);

void main() {
  // WidgetsFlutterBinding.ensureInitialized();
  FxWidgetsFlutterBinding.ensureInitialized(
      uiBlueprints: uiSize, enableLog: true);
  runApp(MyApp());
}
// void main() {
//   return runFxApp(MyApp(),
//       uiBlueprints: uiSize, onEnsureInitialized: () {}, enableLog: true);
// }

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: '设计尺寸${uiSize.toString()}'),
      builder: FxTransitionBuilder(builder: null),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);
  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title ?? ""),
        actions: <Widget>[
          ElevatedButton(
              onPressed: () {
                Navigator.push(context, MaterialPageRoute(builder: (context) {
                  return NextPage();
                }));
              },
              child: Text("NextPage")),
        ],
      ),
      body: SingleChildScrollView(
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              Text(
                'MediaQuery.of(context).size=${MediaQuery.of(context).size}',
              ),
              Text(
                '设计尺寸 414x896 ',
                style: TextStyle(fontSize: 20),
              ),
              Text(
                '设计尺寸 414x896 fontSize: 19',
                style: TextStyle(fontSize: 19),
              ),
              Text(
                '设计尺寸 414x896 fontSize: 18',
                style: TextStyle(fontSize: 18),
              ),
              Row(
                children: <Widget>[
                  SizedBox(
                    width: 100,
                    height: 100,
                    child: Container(
                      color: Colors.orange,
                      child: Text("w= 100,    h=100"),
                    ),
                  ),
                  SizedBox(
                    width: 314,
                    height: 100,
                    child: Container(
                      color: Colors.green,
                      child: Text("w=314"),
                    ),
                  ),
                ],
              ),
              Row(
                children: <Widget>[
                  Container(
                    child: Text("W= 414"),
                    width: 414,
                    height: 80,
                    color: Colors.greenAccent,
                  ),
                ],
              ),
              Row(
                children: <Widget>[
                  Container(
                    child: Text("W= 415"),
                    width: 415,
                    height: 80,
                    color: Colors.lightBlue,
                  ),
                ],
              ),
              Text(
                '设计尺寸 300x510',
              ),
              Row(
                children: <Widget>[
                  Container(
                    child: Text("W= 100"),
                    width: 100,
                    height: 30,
                    color: Colors.lightBlue,
                  ),
                  Container(
                    child: Text("W= 200"),
                    width: 200,
                    height: 30,
                    color: Colors.red,
                  ),
                ],
              ),
              Row(
                children: <Widget>[
                  Container(
                    child: Text("W= 100"),
                    width: 100,
                    height: 30,
                    color: Colors.yellow,
                  ),
                  Container(
                    child: Text("W= 100"),
                    width: 100,
                    height: 30,
                    color: Colors.lightBlue,
                  ),
                  Container(
                    child: Text("W= 100"),
                    width: 100,
                    height: 30,
                    color: Colors.yellow,
                  ),
                ],
              ),
              Row(
                children: <Widget>[
                  Container(
                    child: Text("W= 301"),
                    width: 300,
                    height: 30,
                    color: Colors.cyanAccent,
                  ),
                  Container(
                    child: Text(""),
                    width: 1,
                    height: 30,
                    color: Colors.red,
                  ),
                ],
              ),
              Text(
                '$_counter',
                style: Theme.of(context).textTheme.display1,
              ),
            ],
          ),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

class NextPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    print("@@NextPageMediaQuery ${MediaQuery.of(context)}");
    return Scaffold(
      appBar: AppBar(
        title: Text("next page"),
      ),
      backgroundColor: Colors.green,
      body: Container(
//        margin: EdgeInsets.only(top: MediaQuery.of(context).padding.top),
        height: Info.instance.actualDpSize.height,
        color: Colors.red,
        child: Column(
          children: <Widget>[
            Container(
              height: Info.instance.deltaLength.abs(),
              width: double.infinity,
              color: Colors.deepPurpleAccent,
              child: Text("deltaHeight=${Info.instance.deltaLength}"),
            ),
            Text(
                "Info.instance.actualDpSize.height=${Info.instance.actualDpSize}"),
            Text(
              '设计尺寸 414x896',
              style: TextStyle(fontSize: 20),
            ),
            Text(
              '设计尺寸 414x896 fontSize: 19',
              style: TextStyle(fontSize: 19),
            ),
            Text(
              '设计尺寸 414x896 fontSize: 18',
              style: TextStyle(fontSize: 18),
            ),
            Text(
              '设计尺寸 300x510',
            ),
            Row(
              children: <Widget>[
                Container(
                  child: Text("W= 100"),
                  width: 100,
                  height: 30,
                  color: Colors.lightBlue,
                ),
                Container(
                  child: Text("W= 200"),
                  width: 200,
                  height: 30,
                  color: Colors.red,
                ),
              ],
            ),
            Row(
              children: <Widget>[
                Container(
                  child: Text("W= 100"),
                  width: 100,
                  height: 30,
                  color: Colors.yellow,
                ),
                Container(
                  child: Text("W= 100"),
                  width: 100,
                  height: 30,
                  color: Colors.lightBlue,
                ),
                Container(
                  child: Text("W= 100"),
                  width: 100,
                  height: 30,
                  color: Colors.yellow,
                ),
              ],
            ),
            Padding(
              padding: EdgeInsets.only(top: 100),
            ),
            Row(
              children: <Widget>[
                Container(
                  child: Text("W= 301"),
                  width: 300,
                  height: 30,
                  color: Colors.cyanAccent,
                ),
                Container(
                  child: Text(""),
                  width: 1,
                  height: 30,
                  color: Colors.red,
                ),
              ],
            ),
            RaisedButton(
              onPressed: () {
                Navigator.push(context, MaterialPageRoute(builder: (context) {
                  return NextPage2();
                }));
              },
              child: Text("NextPage"),
            ),
            RaisedButton(
              onPressed: () {
                Navigator.push(context, MaterialPageRoute(builder: (context) {
                  return NextPage3();
                }));
              },
              child: Text("NextPage3"),
            ),
            RaisedButton(
              onPressed: () {
                Navigator.push(context, MaterialPageRoute(builder: (context) {
                  return NextPage4();
                }));
              },
              child: Text("NextPage4"),
            ),
            RaisedButton(
              onPressed: () {
                Navigator.push(context, MaterialPageRoute(builder: (context) {
                  return RepetitionLine();
                }));
              },
              child: Text("RepetitionLine"),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        tooltip: 'Increment',
        onPressed: () {},
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

class RepetitionLine extends StatelessWidget {
  RepetitionLine({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    double value = 50.6;
    return Scaffold(
      body: Column(
        children: [
          SizedBox(
            height: 300,
            child: Scaffold(
              appBar: AppBar(
                  title: Text('复现线条间隙问题'),
                  elevation: 0,
                  backgroundColor: Color(0xFF007D5F)),
              backgroundColor: Color(0xFF007D5F),
              body: Container(color: Color(0xFF007D5F), height: 200),
            ),
          ),
          SizedBox(height: 100),
          Container(color: Colors.green, height: 50),
          Container(color: Colors.green, height: 50),
          Container(color: Colors.green, height: 50),
          Container(color: Colors.green, height: value),
          Container(color: Colors.green, height: value),
          Container(color: Colors.green, height: value),
          Row(
            children: [
              Container(color: Colors.green, height: 50, width: value),
              Container(color: Colors.green, height: 50, width: value),
              Container(color: Colors.green, height: 50, width: value),
            ],
          )
        ],
      ),
    );
  }
}

class NextPage2 extends StatelessWidget {
  Size size = Size.zero;

  @override
  Widget build(BuildContext context) {
    size = MediaQuery.of(context).size;
    // TODO: implement build
    print("build哈哈哈size=$size");
    return Scaffold(
      // appBar: AppBar(
      //   title: Text("哈哈哈"),
      // ),
      body: Column(
        children: [
          Container(
            color: Colors.green,
            width: double.infinity,
            height: MediaQuery.of(context).padding.top,
          ),
//           Container(
//               width: double.infinity,
//               height: MediaQuery.of(context).padding.top,
//               color: Colors.red),
          Container(
              color: Colors.deepPurpleAccent,
              width: double.infinity,
              height: Info.instance.bodyMaxLength,
              // height: Info.instance.uiBlueprints.length+Info.instance.deltaLength,
              // height: double.infinity,
              // height:size.height,
              child: Text(
                  "body${Info.instance.bodyMaxLength} kToolbarHeight=$kToolbarHeight MediaQuery.of(context).padding.top=${MediaQuery.of(context).padding.top}")),
        ],
      ),
    );
  }
}

class NextPage3 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return Scaffold(
      body: Column(
        children: [
          Container(
            width: double.infinity,
            height: Info.instance.uiBlueprints.length,
            color: Colors.red,
            child: Text("PinTenonWidget测试"),
          ),
          PinTenonWidget(),
        ],
      ),
    );
  }
}

class NextPage4 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return Scaffold(
      appBar: AppBar(
        title: Text("PinTenonWidget测试"),
      ),
      body: Column(
        children: [
          Container(
            width: double.infinity,
//            margin: EdgeInsets.only(top: MediaQuery.of(context).padding.vertical),
            height: Info.instance.bodyMaxLength - kToolbarHeight,
            color: Colors.red,
            child: Text("${_getText()}"
                "PinTenonWidget测试 ${MediaQuery.of(context).padding.vertical}"),
          ),
        ],
      ),
    );
  }

  String _getText() {
    StringBuffer sb = new StringBuffer();
    sb.write("int: ${getT<int>()} ${1.runtimeType.toString()} \n");
    sb.write("String: ${getT<String>()} ${"".runtimeType.toString()} \n");
    sb.write(
        "dynamic: ${getT<dynamic>()} ${dynamic.runtimeType.toString()} \n");
    sb.write("List<String>: ${getT<List<String>>()} ${[
      ""
    ].runtimeType.toString()} \n");
    sb.write(
        "List<String>: ${getT<List<String>>()} ${<String>[].runtimeType.toString()} \n");
    sb.write("bool: ${getT<bool>()} ${true.runtimeType.toString()} \n");
    return "${sb.toString()}";
  }

  getT<T>() {
    return T.runtimeType.toString();
  }
}
